tensorboard/webapp/persistent_settings/_data_source/persistent_settings_data_source.ts

```diff
@@ -0,0 +1,121 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {Injectable} from '@angular/core';
+import {EMPTY, Observable, of} from 'rxjs';
+import {map, tap} from 'rxjs/operators';
+import {LocalStorage} from '../../util/local_storage';
+import {BackendSettings, PersistableSettings} from './types';
+
+const LEGACY_METRICS_LOCAL_STORAGE_KEY = '_tb_global_settings.timeseries';
+const GLOBAL_LOCAL_STORAGE_KEY = '_tb_global_settings';
+
+@Injectable()
+export abstract class PersistentSettingsDataSource {
+  abstract setSettings(
+    partialSetting: Partial<PersistableSettings>
+  ): Observable<void>;
+  abstract getSettings(): Observable<Partial<PersistableSettings>>;
+}
+
+/**
+ * An implementation of PersistentSettingsDataSource that stores global settings
+ * in browser local storage.
+ */
+@Injectable()
+export class PersistentSettingsDataSourceImpl
+  implements PersistentSettingsDataSource {
+  constructor(private readonly localStorage: LocalStorage) {}
+
+  private serializeSettings(settings: Partial<PersistableSettings>): string {
+    const serializableSettings: BackendSettings = {
```

Comment:
A lot of neat stuff going on in this PR!  I have a couple high level questions:
- How would an internal-only TensorBoard, or something like TensorBoard.dev define persistent settings?  Wdyt about losing type safety of `BackendSettings`, and instead of hardcoding metrics-specific settings here, just treating it as a bag of key/values?
- There is already a 'Settings' feature/module.  Would it make sense to eventually merge the 'PersistentSettings' feature/module, or is the plan to maintain both?

Also, thanks for migrating the legacy settings to the new scheme in this PR!

Reply:
> How would an internal-only TensorBoard, or something like TensorBoard.dev define persistent settings? Wdyt about losing type safety of BackendSettings, and instead of hardcoding metrics-specific settings here, just treating it as a bag of key/values?

Losing type safety is not great so I took an approach to make use of generics and custom converter instead. Simple key-value will surely bite us when properties get mangled since internal types in metrics will get mangled and will not map nicely onto unmangled key-value settings (and at some point/layer, we need to do a translation between unmangled <-> mangled).

> There is already a 'Settings' feature/module. Would it make sense to eventually merge the 'PersistentSettings' feature/module, or is the plan to maintain both?

Plan is to maintain both. Putting persistence logic inside the settings made it quite busy and it does not pain me much to have a separate folder for "featureless" feature.

Reply:
Thanks for the explanation. `Persisted settings` and `Settings` are both called "settings". How do you plan to make it clearly that they're different if the plan is to maintain both? 
Can we change `Persisted settings` or `Setting` to `experiment configuration`, which sounds more intuitive to me?

Reply:
> Can we change Persisted settings or Setting to experiment configuration, which sounds more intuitive to me?

The persisted_settings module is only responsible with persisting the data (I was thinking about calling it "User preference" but an obvious question someone will ask is, '"Preference" vs. "Settings": they are synonym! why?'. This is why I stuck with "settings". I can leave some module doc to explain the difference for now.